plugin/src/main/java/io/jenkins/plugins/remotingkafka/KafkaSecretManager.java

```diff
@@ -0,0 +1,100 @@
+    private static final Logger LOGGER = Logger.getLogger(KafkaSecretManager.class.getName());
+    private static final Charset UTF_8 = Charset.forName("UTF-8");
+    private static final HMACConfidentialKey AGENT_SECRET =
+            new HMACConfidentialKey(KafkaSecretManager.class, "secret");
+    private final String agentName;
+    private final String producerTopic;
+    private final String producerKey;
+    private final String consumerTopic;
+    private final String consumerKey;
+    private final int producerPartition;
+    private final int consumerPartition;
+    private Producer<String, byte[]> producer;
+    private Consumer<String, byte[]> consumer;
+    /**
+     * Timeout in milliseconds.
+     */
+    private int timeout;
+
+    public KafkaSecretManager(String agentName, KafkaTransportBuilder settings, int timeout) {
+        this.agentName = agentName;
+        this.producer = settings.getProducer();
+        this.consumer = settings.getConsumer();
+        this.producerTopic = settings.getProducerTopic();
+        this.producerKey = settings.getProducerKey();
+        this.producerPartition = settings.getProducerPartition();
+        this.consumerTopic = settings.getConsumerTopic();
+        this.consumerKey = settings.getConsumerKey();
+        this.consumerPartition = settings.getConsumerPartition();
+        this.timeout = timeout;
+    }
+
+    public static String getConnectionSecret(String agentName) {
+        return AGENT_SECRET.mac(agentName);
+    }
+
+    public boolean waitForValidAgent() throws RemotingKafkaConfigurationException {
+        initHandshake();
+        return waitForSecret();
+    }
+
+    private boolean waitForSecret() {
+        String connectionSecret = getConnectionSecret(agentName);
+        String agentSecret = "";
+        TopicPartition partition = new TopicPartition(consumerTopic, consumerPartition);
+        consumer.assign(Arrays.asList(partition));
+        long start = System.currentTimeMillis();
+        while (true) {
+            long current = System.currentTimeMillis();
+            if (current - start > timeout) {
+                LOGGER.warning("Timeout for waiting a secret from agent, restarting the agent...");
+                consumer.close();
+                return false;
+            }
+            ConsumerRecords<String, byte[]> records = consumer.poll(0);
+            consumer.commitSync();
+            for (ConsumerRecord<String, byte[]> record : records) {
+                String receivedValue = new String(record.value(), UTF_8);
+                LOGGER.info("Received a secret=" + receivedValue);
+                if (record.key().equals(consumerKey)) {
+                    agentSecret = receivedValue;
+                    if (connectionSecret.equals(agentSecret)) {
+                        consumer.close();
+                        return true;
+                    } else {
+                        LOGGER.warning("Please sent a valid secret");
```

Comment:
This logging does not help without referencing agent name.
I would rather write "Rejected wrong secret for agent {0}" or so